ui/apps/platform/src/Containers/Policies/Wizard/Step3/PolicyCriteriaOptions.tsx

```diff
@@ -0,0 +1,188 @@
+import React, { useState, useEffect, useMemo } from 'react';
+import {
+    Button,
+    Divider,
+    Flex,
+    FlexItem,
+    Title,
+    Toolbar,
+    ToolbarContent,
+    ToolbarItem,
+    TreeView,
+    TreeViewDataItem,
+    TreeViewSearch,
+} from '@patternfly/react-core';
+import { PlusCircleIcon } from '@patternfly/react-icons';
+import { kebabCase } from 'lodash';
+import { useFormikContext } from 'formik';
+
+import { Policy } from 'types/policy.proto';
+import { Descriptor } from './policyCriteriaDescriptors';
+import { getEmptyPolicyFieldCard } from '../../policies.utils';
+
+import './PolicyCriteriaOptions.css';
+
+function getKeysByCategory(keys) {
+    const categories = {};
+    keys.forEach((key) => {
+        const { category } = key;
+        if (categories[category]) {
+            categories[category].push(key);
+        } else {
+            categories[category] = [key];
+        }
+    });
+    return categories;
+}
+
+function getPolicyFieldsAsTree(existingGroups, descriptors): TreeViewDataItem[] {
+    const categories = getKeysByCategory(descriptors);
+
+    const treeList = Object.keys(categories).map((category) => ({
+        name: '',
+        title: category,
+        id: kebabCase(category),
+        children: categories[category]
+            .filter((child) => {
+                const alreadyUsed = existingGroups.find(
+                    (group) =>
+                        group.fieldName.toLowerCase() === child?.name?.toLowerCase() ||
+                        group.fieldName.toLowerCase() === child?.label?.toLowerCase()
+                );
+                return !alreadyUsed;
+            })
+            .map<TreeViewDataItem>((child: Descriptor) => ({
+                name: child.longName,
+                title: child.shortName || child.name,
+                id: kebabCase(child.shortName),
+                icon: <PlusCircleIcon />,
+            })),
+    }));
+
+    return treeList;
+}
+
+type PolicyCriteriaOptionsProps = {
+    descriptors: Descriptor[];
+    selectedSectionIndex: number;
+};
+
+function PolicyCriteriaOptions({ descriptors, selectedSectionIndex }: PolicyCriteriaOptionsProps) {
+    // const [activeItems, setActiveItems] = useState<TreeViewDataItem[]>([]);
+    const [allExpanded, setAllExpanded] = useState(false);
+    const [filteredItems, setFilteredItems] = useState<TreeViewDataItem[]>([]);
+    const [isFiltered, setIsFiltered] = useState(false);
+    const { values, setFieldValue } = useFormikContext<Policy>();
+    const { policyGroups } = values.policySections[selectedSectionIndex];
+
+    const selectedSection = values.policySections[selectedSectionIndex];
+
+    const treeDataItems = useMemo(() => getPolicyFieldsAsTree([], descriptors), [descriptors]);
+
+    useEffect(() => {
+        setFilteredItems(treeDataItems);
+        setIsFiltered(false);
+    }, [treeDataItems]);
```

Comment:
Potential code simplification nit: 

It looks like `descriptors` is derived from a call to `getPolicyDescriptors` in the parent's render, so the reference to the prop here will always be different.

Since the `useMemo` would recalculate on each render anyway we could remove the `useMemo` call and calculate `treeDataItems` directly.

Alternatively, I bet we could maintain the string value used in `onSearch` lifted up to the parent's state and then pass that value down with `descriptors`. The `filteredItems` could then be directly calculated in render from `descriptors` + `searchFilterString` and remove both pieces of filter state in this component, the `useMemo` call, and the `useEffect`.

We would have to benchmark it, but I imagine the limited items returned in the policy field tree take roughly the same time to compute as the second render caused by the `useEffect` and IMO it would be easier to understand if calculated directly.

Reply:
I spent some time trying to iron out all the data manipulations needed when I lifted the state up. The biggest challenge is that the descriptor object is needed to add a new criterion from this component, but the nested TreeViewDataItems aren't in the correct format.

I tried extending the TreeViewDataItem items to include the original descriptor properties needed to add it as a criteria, but it seemed more complex than just repeating the needed transformation in this component.